### ISSUE TITLE
Implementing IEnumerator<unit> seems impossible

### ISSUE THREAD

**Issue by NinoFloris** [dotnet/fsharp#9913]
Running into this again (for a usecase in Ply)

```fs
open System.Collections.Generic

[<Struct>]
type PlyEnumerator = 
    interface IEnumerator<unit> with
        member _.MoveNext() = true
        member _.Reset() = ()
        member _.Dispose() = ()
        member _.Current: obj = () :> _
        member _.Current: unit = ()
```

```
    error FS0358: The override for 'get_Current : unit -> unit' was ambiguous
    error FS3213: The member 'get_Current : unit -> unit' matches multiple overloads of the same method.
    Please restrict it to one of the following:

       get_Current : unit -> unit

       get_Current : unit -> obj.
```

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzgHwgAcYA7AAgGUBPXAFxgFsA6AYQgwxjDoEsJSuZgHEyMKLzABYAFABtADyU6UAK48AfAF1ZdaiXIAFDNQCipVY3EBDOtHIBecrPKvyvUgyjZrYGOQBJc0sbOygFVVJeOg1yAHdogAsXN1SrRmBxcgB9ZgBZCAA3GAA5GAQ6AAoASkdyFVUYFNTXdMyoHOYAJRhcGCrapxrmlras3IARXlwiCD6auuGZFrcxjtzWVShYTxByCGAAK0XakFjskbSmds7N7bI6Pcjok9kgA==

In general (not just IEnumerator) F# acts weird around interfaces that are implemented at unit, the same would work fine if you would create your own unit type. Something to do with the conversion  to void that is normally tried?

like so
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzgHwgAcYA7AAgGUBPXAFxgFsA6AYQgwxjDoEsJSuZgHEyMKLzABYAFCy61EuQCqpXnXIBeFWrqzZAbQA8lOlACuPAHwBdeYpjkAChmoBRUucbiAhnWha5LLkIeS8pAxQ2D5gjgCSHl6+/lBGqupW5ADu6gAWwaGF3ozA4uQA+swAshAAbjAAcjAIdAAUAJSBZuYwBYUhxaVQFcwASjC4MG2d2h19/YNllQAivLhEEJMdgXMy/aGLw5Ws5lCwESDkEMAAVoHpGiCZ5fNFTEMjJ2dkdJcP97pZEA==

EDIT: `IEnumerator<unit>` seems to be a special case of `Interface<unit>`, it just does not surface the same error that could have lead me to the issues mentioned below.

**Comment by NinoFloris** [comment]
I see there is some background in these issues, but I won't pretend to get the explanations. 

https://github.com/dotnet/fsharp/pull/1377
https://github.com/dotnet/fsharp/issues/35

Reading between the lines it comes down to:
when unit is encountered in return argument position in a member definition we substitute void for unit. When we do this substitution such that it violates the contract to be implemented we return a compiler error today.

Bypassing this is possible by carrying the type parameter into your implementation, as *instantiations* at unit will work (as can be seen in this [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzgHwgAcYA7AAgGUBPXAFxgFsA6AYQgwxjDoEsJSuZgHEyMKLzABYAFCyA2gB5KdKAFceAPgC6sutRLkAChmoBRUmsbiAhnWiKA5HU3kAvOVnlv5XqQZQ2DZgMOQAkhZWtvZQTi7kAO68dAAWXj4Z1ozA4uQA+swAshAAbjAAcjAIdAAUAJTu5KpqMOkZ3lk5UPnMAEowuDC1DR71be2duQUAIry4RBCD9Y1jMu0+k90FrGpQsP4g5BDAAFaNAKqkYCncANYwACbMDzBBahj22Ip5riCueeNMkwuj0dnsyHRDs4LlcbmB7k8Xm8PhAvj9AT5ZFw6OQEMsPKQYAljKZItYoHYHGpSMlNMtZEA=)).

Simply put the current implementation is flawed and should change to take unit into account at contractual member definitions.

**Comment by cartermp** [comment]
Certainly an inconsistency here. Arguably a bug? @dsyme should be the judge of that

**Comment by dsyme** [comment]
Yes, this has always been a known design limitation.  It should have been documented in the language spec however  I've searched for it and can't see it.

For non-struct implementations you can workaround like this:
```
open System.Collections
open System.Collections.Generic

type PlyEnumerator<'T>(t: 'T) = 
    interface IEnumerator<'T> with
        member _.Dispose() = ()
        member _.Current = t
    interface IEnumerator with
        member _.MoveNext() = true
        member _.Reset() = ()
        member _.Current: obj = () :> _

type PlyEnumerator() = 
   inherit PlyEnumerator<unit>( () )
```

I don't think there is a workaround for struct implementations.

**Comment by dsyme** [comment]
I'll close this as by-design for F# 6 - please add a language suggestion if you like
